### PR TITLE
Reference implementation of BIP135 (generalized versionbits)

### PR DESCRIPTION
**Pull request by sanch0panza**
This is an initial reference implementation of my 'generalized versionbits' BIP proposal, the draft specification of which you can find at this link:
https://github.com/bitcoin/bips/blob/master/bip-0135.mediawiki

It provides flexibly configurable parameters for each of the 29 BIP9 deployment bits, including the possibility of

- different thresholds / window sizes
- customizable grace periods (based either on minimum number of blocks or minimum time in seconds elapsed since lock-in)
- explicitly agnostic to fork type (not only intended for soft forks)

The configuration of the deployment bits can be read in from a comma-separated values file, and using the `-dumpforks` option, the built-in configuration can also be dumped out in CSV format to standard output.

BIP135 aims to be completely backward compatible (if suitably configured) with the BIP9 state machine. I have taken care to ensure this by leaving the original tests as unchanged as possible, and adding tests for the new functionality in separate scripts / programs.

The "unknown versions" alerting logic in main.cpp needed to be replaced, which slightly alters its behaviour. It now alerts based on a flat > 50/100 threshold and window, and does not attempt to divine the activation states of unknown signals. It also reports on first detection of an unknown signal, as well as gradated (25%, 50%, 70%, 90%, 95%) crossing thresholds, and loss of previously detected unknown signals during a 100-block window.
The 'unknown version' check itself - performed against the computed version according to Unlimited code - should be backward compatible. The notification / alerting code has only been lightly tested, as it is basically optional (not seen as part of core BIP135 requirements) and probably should be implemented in a fancier way, to allow more user control over when/what/how it reports.

In some places this reference implementation is surely a little worse technically than it should be, since I'm not an expert C++ programmer. Please feel free to criticize and offer suggestions for improvement. I'll do my best to provide rationale about any aspects of the implementation.

---

Update: This is a fresh PR created based on https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/458 which experienced problems updating in github while attempting to rebase on BU 1.3.0.x.

### CODE REVIEW

**Comment by sickpig** [comment]
thanks for doing that ! 

/cc @gandrewstone

**Comment by sanch0panza** [comment]
I need to check the basic regression tests for impacts by this PR. Could be that changes on the base coupled with these changes leads to some failing basic tests. I recall on the previous rebase prior to 1.3.0.x the basic tests still passed.

Before a merge I would also like to get bip9-softforks test to pass, which it doesn't seem to do anymore. 
One of the key points of this PR is to remain compatible with BIP9. Not sure yet what changed, but the test framework is getting a NoneType where it expects an integer.

**Comment by sickpig** [comment]
the travis failure on `make check` seems to be unrelated the changes included in the PR. 

same happens for @awemany's #1087. 

seems we have some kind of problem w/ travis and the cpp tests suite. 

trying to reproduce locally.

**Comment by sanch0panza** [comment]
my test results for basic suite right now

```
TEST                           | PASSED | DURATION

tweak.py                       | True   | 3 s
may152018_forkactivation_2.py  | False  | 31 s
validateblocktemplate.py       | False  | 36 s
may152018_forkactivation_1.py  | True   | 69 s
notify.py                      | True   | 69 s
bip68-112-113-p2p.py           | True   | 161 s
listtransactions.py            | True   | 161 s
receivedby.py                  | False  | 63 s
excessive.py                   | False  | 330 s
mempool_resurrect_test.py      | True   | 33 s
txn_doublespend.py --mineblock | False  | 68 s
txn_clone.py                   | False  | 45 s
wallet.py                      | True   | 414 s
getchaintips.py                | True   | 95 s
mempool_spendcoinbase.py       | True   | 33 s
rest.py                        | True   | 127 s
rawtransactions.py             | True   | 153 s
httpbasics.py                  | True   | 14 s
mempool_reorg.py               | False  | 37 s
multi_rpc.py                   | True   | 15 s
proxy_test.py                  | True   | 14 s
parallel.py                    | True   | 628 s
signrawtransactions.py         | True   | 36 s
zapwallettxes.py               | True   | 130 s
nodehandling.py                | True   | 24 s
merkle_blocks.py               | True   | 147 s
decodescript.py                | True   | 36 s
blockchain.py                  | False  | 38 s
disablewallet.py               | True   | 3 s
fundrawtransaction.py          | False  | 238 s
keypool.py                     | True   | 21 s
invalidtxrequest.py            | True   | 43 s
reindex.py                     | True   | 191 s
abandonconflict.py             | False  | 37 s
p2p-versionbits-warning.py     | False  | 67 s
importprunedfunds.py           | True   | 86 s
sendheaders.py                 | True   | 210 s
walletbackup.py                | True   | 420 s
thinblocks.py                  | True   | 118 s

ALL                            | False  | 4444 s (accumulated)
```

**Comment by sickpig** [comment]
@sanch0panza was referring to this error in the unit tests

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/380131087#L1506

but indeed we need to fix also the ones we find in the functional python tests. 

speaking of which `may152018_forkactivation_2.py` is unrelated to your PR, it has to do w/ the fact that we passed the protocol upgrade date so we need to change it in any case. 

`mempool_reorg.py`, `excessive.py` failing also on current `dev` w/o your changes applied on top and I guess the reason are similar to the one that cause `may152018_forkactivation_2` failure. 

So the "net" failures of this PR should be:

```
validateblocktemplate.py       | False  | 36 s
receivedby.py                  | False  | 63 s
txn_doublespend.py --mineblock | False  | 68 s
txn_clone.py                   | False  | 45 s
blockchain.py                  | False  | 38 s
fundrawtransaction.py          | False  | 238 s
abandonconflict.py             | False  | 37 s
p2p-versionbits-warning.py     | False  | 67 s
```

**Comment by sanch0panza** [comment]
@sickpig : Thanks for the input. I re-tested the basic suite on `dev` at a655d02dfffbb933496b0f8dd7531ed02f34cc5a .

Two tests failed in my PR above that didn't fail on `dev` in my run: `walletbackup` and `p2p-versionbits-warning`. I ran those two again on the PR branch and now they passed. So they are unstable.

The rest that failed in my PR also failed on `dev` without this PR, and some passed that were failed on `dev` (`getchaintips`, `mempool_spendcoinbase.py`).

Will take time to sort out where there are problems that may be linked to the PR.

**Comment by sanch0panza** [comment]
I think the best approach for me is to try to help fix tests on `dev` as far as possible, then rebase again later, to address any defects that might be due to the PR.

**Comment by sickpig** [comment]
That would be great! Thanks.

File: src/consensus/params.h
Comment (ptschip) [inline src/consensus/params.h:29]: These bits should probably be back to unassigned?  segwit of course was never done here and bip68,112,113 has long ago activated.

File: src/test/forkscsv_tests.cpp
Comment (ptschip) [inline src/test/forkscsv_tests.cpp:109]: maybe add a test using some special characters

File: src/versionbits.cpp
Comment (ptschip) [inline src/versionbits.cpp]: is this reference to segwit still needed?

File: doc/bip-genvbvoting-condensed-requirements.txt
Comment (ptschip) [inline doc/bip-genvbvoting-condensed-requirements.txt]: I think this file should rather have an *.md extention rather than a *.txt  ?

File: qa/rpc-tests/bip135-grace.py
Comment (ptschip) [inline qa/rpc-tests/bip135-grace.py]: if you're going to remove these sleeps then shouldn't you be setting sync_every_block=true...otherwise won't you get  spurious failures on slower systems that don't sync as fast since you're assuming in the next steps that the blocks are syncd?

**Review by ptschip — CHANGES_REQUESTED** [review]
@sanch0panza I've take a first pass over the code and will do some followup testing today and tomorrow.

File: src/test/forkscsv_tests.cpp
Comment (ptschip) [inline src/test/forkscsv_tests.cpp]: I'm getting a compile error here "Boost auto testcase passed 2 args but only 1 expected"...I"m wondering what version of boost you're running when you compile?  I"m on 1.58

Comment (ptschip) [reply src/test/forkscsv_tests.cpp]: I commented out the utf::depends_on and the tests run just fine...not sure why you need that or if it's really need it maybe you can just combine the two tests into one?

**Comment by gandrewstone** [comment]
I'd like to get the spec pulled into the doc directory in the BU repo as part of this PR.  Also here are 2 quotes that may be contradictory so I'd like to see clarification added to the spec:

"The threshold specifies a number of blocks, in the range of 1..windowsize, which must signal for a fork in order to lock it in. ***The support is measured when the chain height is evenly divisible by the windowsize***. If the windowsize is set to 2016 (as in BIP9) this coincides with the 2016-block re-targeting intervals."

"The signaling bits SHALL be tallied whenever the head of the active chain changes (including after reorganizations)."

File: src/consensus/params.h
Comment (sanch0panza) [reply src/consensus/params.h:29]: @ptschip : I agree with you. Will submit a patch to unassign.

File: src/test/forkscsv_tests.cpp
Comment (sanch0panza) [reply src/test/forkscsv_tests.cpp:109]: Good idea

**Comment by sanch0panza** [comment]
@gandrewstone : I'll add a patch to pull in the spec, and to clarify the passages you mention.

They are not contradictory in effect - the first refers to evaluating for a fork lock-in at only those intervals determined by the windowsize. The second refers to incrementing the counts when the chain head changes. These are separate concerns, but the spec should be clearer about it.

File: qa/rpc-tests/bip135-grace.py
Comment (gandrewstone) [inline qa/rpc-tests/bip135-grace.py:298]: The code below is the same as above... cut-paste error?

File: qa/rpc-tests/bip135-grace.py
Comment (gandrewstone) [inline qa/rpc-tests/bip135-grace.py:1]: do not use "-" in python file names.  This means they can't be used interactively:
```
$ python
>>> import bip135-grace
  File "<stdin>", line 1
    import bip135-grace
                 ^
SyntaxError: invalid syntax
```
interactive use is nice for development and debugging.

File: qa/rpc-tests/bip135-grace.py
Comment (gandrewstone) [inline qa/rpc-tests/bip135-grace.py:158]: please convert prints to logging.  This cuts down the size of the already-huge travis output.  At the top:
import logging
logging.basicConfig(format='%(asctime)s.%(levelname)s: %(message)s', level=logging.INFO, stream=sys.stdout)

then:
logging.info("begin test_BIP135GraceConditions test")

File: qa/rpc-tests/bip135-threshold.py
Comment (gandrewstone) [inline qa/rpc-tests/bip135-threshold.py:1]: rename without "-" and use logging in this file too, thx!

File: qa/rpc-tests/bip135-threshold.py
Comment (gandrewstone) [inline qa/rpc-tests/bip135-threshold.py:79]: I would like to see a test that verifies:
 
1. the timeout.  The test should produce threshold-1 blocks in windowsize before timeout, and then produces a block with the bit set after timeout.  This should NOT activate (look for self.nodes[0].setmocktime(time) in other tests to see how to manipulate time in your test).

2. Previously I mentioned confusion as to when activation bits were counted.  If they are counted only when blockheight%windowsize==0, I would like to see a test that produces more than threshold blocks with the bit set, but does NOT activate because these blocks cross a windowsize boundary.

3. What edge conditions did I miss?

We need to test these specification edge conditions in the python code because the python code can easily be used for interoperability testing (between full nodes), whereas the C code is quite locked into a specific implementation.

File: src/bitcoind.cpp
Comment (gandrewstone) [inline src/bitcoind.cpp:22]: double include

File: src/chainparams.cpp
Comment (gandrewstone) [inline src/chainparams.cpp:124]: why did the order get swapped?  It'll be a cleaner diff without swapping

File: src/chainparams.cpp
Comment (gandrewstone) [inline src/chainparams.cpp:417]: order swap

File: src/chainparams.cpp
Comment (gandrewstone) [inline src/chainparams.cpp:511]: range check bit >=0 and < MAX_VERSION_BITS_DEPLOYMENTS, DbgAssert(predicate, return false)  to catch in debug mode but continue in release.

File: src/consensus/params.h
Comment (gandrewstone) [inline src/consensus/params.h:58]: semantically, the "max" should be inclusive, so 28.  "What's the maximum score on this test?", you don't answer "101" for the typical 0-100 test score range.  So what is defined as MAX_VERSION_BITS_DEPLOYMENTS, should really be NUM_VERSION_BITS_DEPLOYMENTS -- the number of version bits is 29

Please change the name if you agree, but I'm not going to block the PR for this triviality...

File: src/chainparams.cpp
Comment (gandrewstone) [inline src/chainparams.cpp:509]: Capitalize first letter of function names:
"IsConfiguredDeployment"

File: src/chainparams.cpp
Comment (gandrewstone) [inline src/chainparams.cpp:527]: extra ;

File: src/chainparams.h
Comment (gandrewstone) [inline src/chainparams.h:73]: add "this can only be used during initialization, since modification of the consensus params is not thread safe" to the comment.

File: src/chainparams.h
Comment (gandrewstone) [inline src/chainparams.h:148]: add same as above to the docs of this function too.

File: src/consensus/params.h
Comment (gandrewstone) [inline src/consensus/params.h:80]: does a negative time have meaning?

File: src/consensus/params.h
Comment (gandrewstone) [reply src/consensus/params.h:58]: Also, what is the relationship between this and VERSIONBITS_NUM_BITS?  If they are the same, please consolidate. 1 idea should have exactly 1 constant to help code maintenance.  Otherwise you could have a case where only 1 of the 2 constants gets changed.

File: src/init.cpp
Comment (gandrewstone) [inline src/init.cpp:947]: If the user configured a fork file but we can't find it, I think that we should exit with an error.  

If the default fork file is not there, its ok to log and continue.

File: src/rpc/blockchain.cpp
Comment (gandrewstone) [inline src/rpc/blockchain.cpp:748]: range check "id"

File: src/rpc/blockchain.cpp
Comment (gandrewstone) [inline src/rpc/blockchain.cpp:808]: are these addtl spaces intentional?

File: src/versionbits.cpp
Comment (gandrewstone) [inline src/versionbits.cpp:17]: remove segwit

File: src/versionbits.cpp
Comment (gandrewstone) [inline src/versionbits.cpp:162]: this assertion is guaranteed true due to the previous if statement

File: src/versionbits.cpp
Comment (gandrewstone) [inline src/versionbits.cpp:166]: check pindexPrev for NULL, if its null, some garbage was put in the fork definition, like a period larger than the current block height.  Return THRESHOLD_DEFINED?
EDIT: Nevermind you check null below

File: qa/rpc-tests/bip135-grace.py
Comment (ftrader) [reply qa/rpc-tests/bip135-grace.py]: Thanks will do

File: src/chainparams.cpp
Comment (ftrader) [reply src/chainparams.cpp]: ok

File: qa/rpc-tests/bip135-grace.py
Comment (sanch0panza) [reply qa/rpc-tests/bip135-grace.py:1]: ok, files will be renamed

File: qa/rpc-tests/bip135-grace.py
Comment (sanch0panza) [reply qa/rpc-tests/bip135-grace.py:158]: will do, thanks

File: qa/rpc-tests/bip135-threshold.py
Comment (sanch0panza) [reply qa/rpc-tests/bip135-threshold.py:1]: ok, np

File: src/bitcoind.cpp
Comment (sanch0panza) [reply src/bitcoind.cpp:22]: thanks, will remove

File: src/chainparams.cpp
Comment (sanch0panza) [reply src/chainparams.cpp:124]: not sure, will align to make diff cleaner

File: src/chainparams.cpp
Comment (sanch0panza) [reply src/chainparams.cpp:417]: ok

File: src/chainparams.cpp
Comment (sanch0panza) [reply src/chainparams.cpp:509]: ok, thanks

**Comment by gandrewstone** [comment]
Let me know when you've addressed the comments -- I want to get this merged in before new conflicts occur.

**Pull request merged**
